include/deal.II/optimization/solver_bfgs.h

```diff
@@ -0,0 +1,327 @@
+ * y^{(k)} &:= g^{(k+1)} - g^{(k)} \\
+ * s^{(k)} &:= x^{(k+1)} - x^{(k)} \\
+ * \rho^{(k)} &:= \frac{1}{y^{(k)} \cdot s^{(k)}}
+ * \f]
+ * for a symmetric positive definite $H$. Limited memory variant is
+ * implemented via the two-loop recursion.
+ *
+ * @author Denis Davydov, 2018
+ */
+template <typename VectorType>
+class SolverBFGS : public Solver<VectorType>
+{
+public:
+  /**
+   * Number type.
+   */
+  typedef typename VectorType::value_type Number;
+
+
+  /**
+   * Standardized data struct to pipe additional data to the solver.
+   */
+  struct AdditionalData
+  {
+    /**
+     * Constructor.
+     */
+    explicit AdditionalData(const unsigned int m_max            = 5,
+                            const Number       reject_tolerance = 1e-4,
+                            const bool         debug_output     = false);
+
+    /**
+     * Maximum history size.
+     */
+    const unsigned int m_max;
+
+    /**
+     * Relative tolerance to reject gradient update.
+     */
+    const Number reject_tolerance;
+
+    /**
+     * Print extra debug output to deallog.
+     */
+    const bool debug_output;
+  };
+
+
+  /**
+   * Constructor.
+   */
+  SolverBFGS(SolverControl &       residual_control,
+             const AdditionalData &data = AdditionalData());
+
+  /**
+   * Virtual destructor.
+   */
+  virtual ~SolverBFGS() = default;
+
+  /**
+   * Solve the problem starting from initial state @p x.
+   *
+   * The function @p compute should update the value of gradient of function (second parameter)
+   * based on the input value of unknown (first parameter) and return
+   * the value of function begin minimized.
+   *
+   * @note The value of unknown is intentionally kept as non-const reference
+   * to allow its modifications when it represent a subspace.
+   */
+  void
+  solve(const std::function<Number(VectorType &x, VectorType &g)> &compute,
+        VectorType &                                               x);
+
+  /**
+   * Connect a slot to perform a custom line-search.
+   */
+  boost::signals2::connection
+  connect_line_search_slot(
+    const std::function<
+      Number(Number &f, VectorType &x, VectorType &g, const VectorType &p)>
+      &slot);
+
+  /**
+   * Connect a slot to perform a custom preconditioning.
+   */
+  boost::signals2::connection
+  connect_preconditioner_slot(
+    const std::function<void(VectorType &                         g,
+                             const FiniteSizeHistory<VectorType> &s,
+                             const FiniteSizeHistory<VectorType> &y)> &slot);
+
+
+protected:
+  /**
+   * Additional data to the solver.
+   */
+  const AdditionalData additional_data;
+
+  /**
+   * Signal used to perform line search.
+   * Given the search direction @p p, current value of unknown @p x,
+   * the gradient @p g and value of function @p f,
+   * return the size of the step $x \leftarrow x + \alpha p$,
+   * and update @p x, @p g and @p f accordingly.
```

Comment:
> Can we replace f by value?

we could, but it would be inconsistent with `x`, `p` and `g`, which are all single letters.

> does the usercode need to perform x=x+alpha p or does the code do this?

this function is responsible, that's why I wrote

> update @p x, @p g and @p f accordingly.